File: leetcode/693binary_number_with_alternating_bits/_693binary_number_with_alternating_bits.py
"""
Given a positive integer, check whether it has alternating bits: namely, 
if two adjacent bits will always have different values.

Example 1:
Input: 5
Output: True
Explanation:
The binary representation of 5 is: 101

Example 2:
Input: 7
Output: False
Explanation:
The binary representation of 7 is: 111.
"""
#my solution:
new = [0]
for i in range(32):#列表法，10 = 2，101 = 5，1010 = 10 10101 = 21，101010 = 42…… pow(2,31)<new32;
    if i%2==0:
        new.append(new[i]*2+1)
    else:
        new.append(new[i]*2)

class Solution(object):
    def hasAlternatingBits(self, n):
        """
        :type n: int
        :rtype: bool
        """
        global new
        return n in new


a = Solution()
print(a.hasAlternatingBits(5))
print(a.hasAlternatingBits(7))

##There are also much good solutions:
class Solution(object):
    def hasAlternatingBits(self, n):
        """
        :type n: int
        :rtype: bool
        """
        return '00'in bin(n) and '11'in bin(n)#so crazy!!!

File: leetcode/693binary_number_with_alternating_bits/693binary_number_with_alternating_bits.c
#include<stdio.h>// add c solution
#include<stdbool.h>

int mtrix[33];
void mtrixM() {
	mtrix[0] = 0;
	int i;
	for (i = 1; i<32; i++) {
		if (i % 2 == 0)
			mtrix[i] = mtrix[i - 1] * 2;
		else mtrix[i] = mtrix[i - 1] * 2 + 1;
	}
}

bool hasAlternatingBits(int n) {
	mtrixM();
	int i = 0;
	for (i = 0; i<32; i++) {
		if (n == mtrix[i])
			return true;
	}
	return false;
}

int main() {
	mtrixM();
	int i;
	for (i = 0; i<10; i++) {
		printf("%d ", mtrix[i]);
	}
}
